src/main/java/org/assertj/core/internal/Maps.java

```diff
@@ -670,15 +678,59 @@ public void assertHasSameSizeAs(AssertionInfo info, Map<?, ?> map, Map<?, ?> oth
     }
     failIfEmpty(keys, keysToLookForIsEmpty(placeholderForErrorMessages));
 
-    Set<K> notFound = new LinkedHashSet<>();
-    Set<K> notExpected = new LinkedHashSet<>();
-
-    compareActualMapAndExpectedKeys(actual, keys, notExpected, notFound);
+    Set<K> notFound = getNotFoundKeys(actual, keys);
+    Set<K> notExpected = getNotExpectedKeys(actual, keys);
 
     if (!notFound.isEmpty() || !notExpected.isEmpty())
       throw failures.failure(info, shouldContainOnlyKeys(actual, keys, notFound, notExpected));
   }
 
+  private static <K> Set<K> getNotFoundKeys(Map<K, ?> actual, K[] keys) {
+    return Stream.of(keys)
+                 .filter(key -> !actual.containsKey(key))
+                 .collect(Collectors.toSet());
+  }
+
+  private static <K> Set<K> getNotExpectedKeys(Map<K, ?> actual, K[] keys) {
+    List<K> expectedKeys = java.util.Arrays.asList(keys);
+    try {
+      Set<K> unexpectedKeys = clone(actual).keySet();
+      expectedKeys.forEach(unexpectedKeys::remove);
+      return unexpectedKeys;
+    } catch (NoSuchMethodException | UnsupportedOperationException e) {
+      // actual cannot be copied or is unmodifiable, falling back to LinkedHashMap
+      Set<K> unexpectedKeys = new LinkedHashMap<>(actual).keySet();
+      expectedKeys.forEach(unexpectedKeys::remove);
+      return unexpectedKeys;
+    }
+  }
+
+  @SuppressWarnings({ "unchecked" })
+  private static <K, V> Map<K, V> clone(Map<K, V> map) throws NoSuchMethodException {
+    if (map instanceof Cloneable) {
+      try {
+        return (Map<K, V>) map.getClass().getMethod("clone").invoke(map);
+      } catch (IllegalAccessException | InvocationTargetException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+    try {
+      @SuppressWarnings("rawtypes")
+      Class<? extends Map> type = map.getClass();
+      try {
+        // try with copying constructor
+        return type.getConstructor(Map.class).newInstance(map);
+      } catch (NoSuchMethodException e) {
+        // try with default constructor
+        Map<K, V> newMap = type.getConstructor().newInstance();
+        newMap.putAll(map);
+        return newMap;
+      }
```

Comment:
With this, I don't see the need of bringing in Objenesis. Please, let me know if you think otherwise.

Reply:
I agree we don't need objenesis for now